pylint/lint/parallel.py

```diff
@@ -71,11 +71,9 @@ def _worker_check_single_file(
     _worker_linter.check_single_file_item(file_item)
     mapreduce_data = defaultdict(list)
     for checker in _worker_linter.get_checkers():
-        try:
-            data = checker.get_map_data()
-        except AttributeError:
-            continue
-        mapreduce_data[checker.name].append(data)
+        data = checker.get_map_data()
+        if data is not None:
+            mapreduce_data[checker.name].append(data)
```

Comment:
I think this is a sign we would need a changelog. If someone's subclass implemented `get_map_data()` and returned None, they need to know there won't be a key in `mapreduce_data` now. Obviously we can spare all this detail, but just a short note to say we promoted the map/reduce functionality, and then at that point go ahead and deprecate MapReduceMixin, and then... I'm starting to worry this is solving a problem that's not really a problem.

I would just do this and be done with it. WDYT?
```diff
diff --git a/pylint/lint/parallel.py b/pylint/lint/parallel.py
index a1ecb7ac..6b927173 100644
--- a/pylint/lint/parallel.py
+++ b/pylint/lint/parallel.py
@@ -13,6 +13,7 @@ from typing import TYPE_CHECKING, Any
 import dill
 
 from pylint import reporters
+from pylint.checkers.mapreduce_checker import MapReduceMixin
 from pylint.lint.utils import _patch_sys_path
 from pylint.message import Message
 from pylint.typing import FileItem
@@ -71,11 +72,9 @@ def _worker_check_single_file(
     _worker_linter.check_single_file_item(file_item)
     mapreduce_data = defaultdict(list)
     for checker in _worker_linter.get_checkers():
-        try:
+        if isinstance(checker, MapReduceMixin):
             data = checker.get_map_data()
-        except AttributeError:
-            continue
-        mapreduce_data[checker.name].append(data)
+            mapreduce_data[checker.name].append(data)
     msgs = _worker_linter.reporter.messages
     assert isinstance(_worker_linter.reporter, reporters.CollectingReporter)
     _worker_linter.reporter.reset()
@@ -120,6 +119,7 @@ def _merge_mapreduce_data(
         if checker.name in collated_map_reduce_data:
             # Assume that if the check has returned map/reduce data that it has the
             # reducer function
+            assert isinstance(checker, MapReduceMixin)
             checker.reduce_map_data(linter, collated_map_reduce_data[checker.name])
 ```

Reply:
I can certainly add a changelog. Note that if you currently return `None` it won't do much since then `reduce_map_data` will just return  `list[None]` which is difficult to reduce then.

That would also work, but for me the underlying issue is the `Mixin` to begin with. It seems to overcomplicate stuff which could be relatively simple. We'll never need `MapReduceMixin` with any other class than `BaseChecker` and I can't really see anything that makes it bad to have to have dummy methods on the base class for these methods.

Reply:
I guess I don't feel this is less complicated, given that we're not deprecating the mixin in this PR. I don't feel too strongly about it either way, though.

Reply:
Would you want me to add the deprecation in this PR?

Reply:
Let's see what other reviewers think. If we really need to unblock main we can always do that separately with just fixing mypy.

Reply:
I'll make a quick PR with some disables. Not being able to run `pre-commit` locally makes it impossible to commit in VSCode without ``--no-verify`` which has been bugging me for 24h now.

Reply:
Added a last commit to show potential deprecation.

@Pierre-Sassoulas What do you think about deprecating this class?